### PR TITLE
Link hover preview

### PR DESCRIPTION
**Pull request by jimrandomh**
Known issues:
- [X] Material-UI's `Tooltip` component determines the position of the tooltip based on the initial size of the tooltip contents, assuming the size won't change, but it starts out as a loading-spinner.
- [X] Only links leading to the `posts.single` route get previews. (Adding support for more routes is straightforward, I just haven't done it yet.)
- [X] No optimization has yet been put into the styling of the preview itself.

Possible issues, needs testing:
- [X] Do new usages of `ContentItemBody` interfere with incremental-rerender in any performance-relevant ways?

### CODE REVIEW

**Comment by jimrandomh** [comment]
The biggest limitation of this right now is it can't handle Overcoming Bias links which redirect back to the sequences, and it can't handle the legacy link format with short URLs (`/lw/<base36 ID>`), because that isn't handled through the Routes table. Leaving those for future work.

**Comment by jimrandomh** [comment]
A problem with this: This makes on-site links in posts (regardless of whether they're getting a post-preview) use the `Link` component, rather than an `<a href>`. But many Sequences posts link to each other using onsite links which can only be server-side rendered, and can't be client-side rendered; ie, they're old-style URLs where we have a redirect implemented server side, but don't have a corresponding entry in the Routes table to make that redirect work client side. So as currently written, those links will break.

**Comment by Discordius** [comment]
At least one of the deepscan issues seems real

**Comment by jimrandomh** [comment]
It's not real, in the sense of corresponding to a bug. Will fix anyways for lint's sake though.

**Comment by Discordius** [comment]
Bug: Footnote links show a preview of the current post, which is definitely not the expected behavior. Probably don't show preview for the current post? (Or maybe deactivate it for footnotes)

**Comment by Discordius** [comment]
On this post, the link to the absurdity heuristic doesn't render a preview: 

http://localhost:3000/s/5g5TkQTe9rmPS5vvM/p/R8cpqD3NA4rZxRdQ4

**Comment by jimrandomh** [comment]
Looks like footnotes (as done in the sequences) give their anchor an `id` and use that for the back-link, which is currently not getting passed through.

File: packages/lesswrong/components/common/ContentItemBody.jsx [hunk 1]
```diff
@@ -181,6 +194,28 @@ class ContentItemBody extends Component {
     updateScrollIndicatorClasses();
     block.onscroll = (ev) => updateScrollIndicatorClasses();
   };
+  
+  markHoverableLinks = () => {
+    //if(!Meteor.isServer && this.bodyRef && this.bodyRef.current) {
```
Comment (Raemon) [inline packages/lesswrong/components/common/ContentItemBody.jsx]: remove unused line

File: packages/lesswrong/components/common/ContentItemBody.jsx [hunk 2]
```diff
@@ -181,6 +194,28 @@ class ContentItemBody extends Component {
...
+    if(this.bodyRef && this.bodyRef.current) {
```
Comment (Raemon) [inline packages/lesswrong/components/common/ContentItemBody.jsx]:
```suggestion
    if(this.bodyRef?.current) {
```

**Comment by Raemon** [comment]
I think the domain name whitelist is fairly important – it should at least work for:

alignmentforum.org
alignment-forum.com
lesserwrong.com
lessestwrong.com

(and if it can trivially work with localhost:3000, I think that'd be pretty useful for testing)

File: packages/lesswrong/components/linkPreview/HoverPreviewLink.jsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  };
+};
+
+const HoverPreviewLink = ({innerHTML, href}) => {
+  // FIXME: This is currently client-side-only because 'URL' is a browser-API
+  // class. See the workaround for the same issue in PostsPage.
+  const location = useLocation();
+  
+  // Within-page relative link?
+  if (href.startsWith("#")) {
```
Comment (Raemon) [inline packages/lesswrong/components/linkPreview/HoverPreviewLink.jsx]:
```suggestion
  if (href?.startsWith("#")) {
```

Comment (Raemon) [reply packages/lesswrong/components/linkPreview/HoverPreviewLink.jsx]: (this crashed on my local server when I had a fake or bad link in a post)

Comment (Raemon) [reply packages/lesswrong/components/linkPreview/HoverPreviewLink.jsx]: (might want to add more defensive programming in the various dot-donation variables)

**Comment by Raemon** [comment]
I've reverted the change the PostsItemTooltip, so that it (as originally intended) is hidden on small displays. 

I plan (in the near future) to replace the entire bit where this thing uses the PostsItemTooltip. I'm leaving the other optional updates to the tooltip for now but will probably remove them in an upcoming patch.

**Pull request merged**